Source/WebKit/UIProcess/Extensions/WebExtension.cpp

```diff
@@ -245,6 +245,61 @@ void WebExtension::populateWebAccessibleResourcesIfNeeded()
         parseWebAccessibleResourcesVersion2();
 }
 
+URL WebExtension::resourceFileURLForPath(const String& originalPath)
+{
+    ASSERT(originalPath);
+
+    String path = originalPath;
+
+    if (path.startsWith('/'))
+        path = path.substring(1);
+
+    if (!path.length() || m_resourceBaseURL.isNull())
+        return { };
+
+    auto resourceURL = URL(m_resourceBaseURL, path).fileSystemPath();
+
+    // Don't allow escaping the base URL with "../".
+    if (!resourceURL.startsWith(m_resourceBaseURL.string())) {
+        RELEASE_LOG_ERROR(Extensions, "Resource URL path escape attempt: %s", resourceURL.utf8().data());
+        return { };
+    }
+
+    return URL(resourceURL);
+}
+
+String WebExtension::resourceStringForPath(const String& unsafePath, RefPtr<API::Error>& outError, CacheResult cacheResult, SuppressNotFoundErrors suppressErrors)
+{
+    ASSERT(unsafePath);
+
+    String path = unsafePath;
+
+    // Remove leading slash to normalize the path for lookup/storage in the cache dictionary.
+    if (path.startsWith('/'))
+        path = path.substring(1);
+
+    if (auto cachedData = m_resources.getOptional(path)) {
+        if (auto cachedString = String::fromUTF8(*cachedData); !cachedString.isEmpty())
+            return cachedString;
+    }
+
+    if ((path == generatedBackgroundPageFilename) || (path == generatedBackgroundServiceWorkerFilename))
+        return generatedBackgroundContent();
+
+    auto data = resourceDataForPath(path, outError, CacheResult::No, suppressErrors);
+    auto resourceString = emptyString();
+    if (!data.empty())
+        resourceString = String(data);
```

Comment:
This constructor will use latin-1, so this is worse. Use `String::fromUTF8WithLatin1Fallback()`.

Reply:
Yeah, I know it's not as robust but unless I'm missing something WTF doesn't have support for others